// @firebase/eslint-plugin-security-rules/dist/index.js
import {
  parser_exports
} from "./chunk-Y4KWYCZJ.js";

// src/rules/no-open-reads.ts
var no_open_reads_default = {
  meta: {
    type: "problem",
    docs: {
      description: "disallow 'if true' for read permissions which renders the database readable to the internet.",
      recommended: true
    },
    messages: {
      openRead: "Insecure open read rule detected. Warn on open reads in production."
    },
    schema: []
  },
  create(context) {
    return {
      "Program"(node) {
        const sourceCode = context.sourceCode;
        const services = sourceCode.parserServices || {};
        if (!services.tree) {
          return;
        }
        function walk(antlrNode, callback) {
          if (!antlrNode) return;
          callback(antlrNode);
          if (antlrNode.children) {
            for (const child of antlrNode.children) {
              walk(child, callback);
            }
          }
        }
        walk(services.tree, (antlrNode) => {
          if (antlrNode.constructor && antlrNode.constructor.name === "PermissionDeclarationContext") {
            const text = antlrNode.getText();
            if (text.match(/^allow.*read.*:iftrue;?$/i)) {
              const startToken = antlrNode.start;
              const endToken = antlrNode.stop;
              context.report({
                node,
                loc: {
                  start: { line: startToken.line, column: startToken.column },
                  end: { line: endToken.line, column: endToken.column + 1 }
                },
                messageId: "openRead"
              });
            }
          }
        });
      }
    };
  }
};

// src/rules/no-open-writes.ts
var no_open_writes_default = {
  meta: {
    type: "problem",
    docs: {
      description: "disallow 'if true' for write permissions which renders the database writable to the internet.",
      recommended: true
    },
    messages: {
      openWrite: "Insecure open write rule detected. Error on open writes in production."
    },
    schema: []
  },
  create(context) {
    return {
      "Program"(node) {
        if (!context.filename.endsWith("firestore.rules")) {
          return;
        }
        const sourceCode = context.sourceCode || context.getSourceCode();
        const services = context.parserServices || sourceCode.parserServices || {};
        if (!services.tree) {
          return;
        }
        function walk(antlrNode, callback) {
          if (!antlrNode) return;
          callback(antlrNode);
          if (antlrNode.children) {
            for (const child of antlrNode.children) {
              walk(child, callback);
            }
          }
        }
        walk(services.tree, (antlrNode) => {
          if (antlrNode.constructor && antlrNode.constructor.name === "PermissionDeclarationContext") {
            const text = antlrNode.getText();
            if (text.match(/^allow.*write.*:iftrue;?$/i)) {
              const startToken = antlrNode.start;
              const endToken = antlrNode.stop;
              context.report({
                node,
                loc: {
                  start: { line: startToken.line, column: startToken.column },
                  end: { line: endToken.line, column: endToken.column + 1 }
                },
                messageId: "openWrite"
              });
            }
          }
        });
      }
    };
  }
};

// src/rules/no-redundant-matches.ts
var no_redundant_matches_default = {
  meta: {
    type: "suggestion",
    docs: {
      description: "disallow identical match statements in the same scope",
      recommended: true
    },
    messages: {
      redundantMatch: "Redundant match path '{{pathText}}' overlaps with previously declared match."
    },
    schema: []
  },
  create(context) {
    return {
      "Program"(node) {
        if (!context.filename.endsWith("firestore.rules")) {
          return;
        }
        const sourceCode = context.sourceCode;
        const services = sourceCode.parserServices || {};
        if (!services.tree) {
          return;
        }
        function walk(antlrNode, callback) {
          if (!antlrNode) return;
          callback(antlrNode);
          if (antlrNode.children) {
            for (const child of antlrNode.children) {
              walk(child, callback);
            }
          }
        }
        const scopeMap = /* @__PURE__ */ new Map();
        walk(services.tree, (antlrNode) => {
          if (antlrNode.constructor && antlrNode.constructor.name === "MatchRuleDeclarationContext") {
            const pathDecl = antlrNode.pathDecl ? antlrNode.pathDecl() : null;
            if (!pathDecl) return;
            const pathText = pathDecl.getText();
            const normalizedPath = pathText.replace(/\{[^}=]+(\=\*\*)?\}/g, "{*}");
            let scope = antlrNode.parentCtx;
            while (scope && scope.constructor && !["MatchRuleDeclarationContext", "ServiceDeclarationContext"].includes(scope.constructor.name)) {
              scope = scope.parentCtx;
            }
            if (!scope) return;
            let seenMatches = scopeMap.get(scope);
            if (!seenMatches) {
              seenMatches = /* @__PURE__ */ new Set();
              scopeMap.set(scope, seenMatches);
            }
            if (seenMatches.has(normalizedPath)) {
              const startToken = pathDecl.start;
              const endToken = pathDecl.stop;
              context.report({
                node,
                loc: {
                  start: { line: startToken.line, column: startToken.column },
                  end: {
                    line: endToken.line,
                    column: endToken.column + pathDecl.getText().length
                  }
                },
                messageId: "redundantMatch",
                data: {
                  pathText
                }
              });
            } else {
              seenMatches.add(normalizedPath);
            }
          }
        });
      }
    };
  }
};

// src/index.ts
var rules = {
  "no-open-reads": no_open_reads_default,
  "no-open-writes": no_open_writes_default,
  "no-redundant-matches": no_redundant_matches_default
};
var plugin = {
  meta: {
    name: "@firebase/eslint-plugin-security-rules",
    version: "1.0.0"
  },
  rules
};
var flatRecommended = {
  files: ["**/*.rules"],
  plugins: { "@firebase/security-rules": plugin },
  languageOptions: {
    parser: parser_exports
  },
  rules: {
    "@firebase/security-rules/no-open-reads": "warn",
    "@firebase/security-rules/no-open-writes": "error",
    "@firebase/security-rules/no-redundant-matches": "error"
  }
};
var legacyRecommended = {
  plugins: ["@firebase/security-rules"],
  // Old string array style
  rules: {
    "@firebase/security-rules/no-open-reads": "warn",
    "@firebase/security-rules/no-open-writes": "error",
    "@firebase/security-rules/no-redundant-matches": "error"
  }
};
var configs = {
  "flat/recommended": flatRecommended,
  "recommended": legacyRecommended
};
var index_default = {
  configs,
  parser: parser_exports,
  ...plugin
};
export {
  index_default as default,
  flatRecommended,
  rules
};
