// @firebase/eslint-plugin-security-rules/grammar/FirebaseRulesParserListener.ts
// @ts-nocheck
/**
 * @license
 * Copyright 2026 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

// Generated from grammar/FirebaseRulesParser.g4 by ANTLR 4.13.1
// jshint ignore: start
import antlr4_lib from 'antlr4';
const antlr4 = antlr4_lib.tree ? antlr4_lib : antlr4_lib.default;

// This class defines a complete listener for a parse tree produced by FirebaseRulesParser.
export default class FirebaseRulesParserListener extends antlr4.tree.ParseTreeListener {

	// Enter a parse tree produced by FirebaseRulesParser#ruleset.
	enterRuleset(ctx) {
	}

	// Exit a parse tree produced by FirebaseRulesParser#ruleset.
	exitRuleset(ctx) {
	}


	// Enter a parse tree produced by FirebaseRulesParser#versionStatement.
	enterVersionStatement(ctx) {
	}

	// Exit a parse tree produced by FirebaseRulesParser#versionStatement.
	exitVersionStatement(ctx) {
	}


	// Enter a parse tree produced by FirebaseRulesParser#importStatement.
	enterImportStatement(ctx) {
	}

	// Exit a parse tree produced by FirebaseRulesParser#importStatement.
	exitImportStatement(ctx) {
	}


	// Enter a parse tree produced by FirebaseRulesParser#rulesetStatement.
	enterRulesetStatement(ctx) {
	}

	// Exit a parse tree produced by FirebaseRulesParser#rulesetStatement.
	exitRulesetStatement(ctx) {
	}


	// Enter a parse tree produced by FirebaseRulesParser#serviceDeclaration.
	enterServiceDeclaration(ctx) {
	}

	// Exit a parse tree produced by FirebaseRulesParser#serviceDeclaration.
	exitServiceDeclaration(ctx) {
	}


	// Enter a parse tree produced by FirebaseRulesParser#serviceStatement.
	enterServiceStatement(ctx) {
	}

	// Exit a parse tree produced by FirebaseRulesParser#serviceStatement.
	exitServiceStatement(ctx) {
	}


	// Enter a parse tree produced by FirebaseRulesParser#functionDeclaration.
	enterFunctionDeclaration(ctx) {
	}

	// Exit a parse tree produced by FirebaseRulesParser#functionDeclaration.
	exitFunctionDeclaration(ctx) {
	}


	// Enter a parse tree produced by FirebaseRulesParser#bindingDeclaration.
	enterBindingDeclaration(ctx) {
	}

	// Exit a parse tree produced by FirebaseRulesParser#bindingDeclaration.
	exitBindingDeclaration(ctx) {
	}


	// Enter a parse tree produced by FirebaseRulesParser#functionSignature.
	enterFunctionSignature(ctx) {
	}

	// Exit a parse tree produced by FirebaseRulesParser#functionSignature.
	exitFunctionSignature(ctx) {
	}


	// Enter a parse tree produced by FirebaseRulesParser#paramList.
	enterParamList(ctx) {
	}

	// Exit a parse tree produced by FirebaseRulesParser#paramList.
	exitParamList(ctx) {
	}


	// Enter a parse tree produced by FirebaseRulesParser#functionBody.
	enterFunctionBody(ctx) {
	}

	// Exit a parse tree produced by FirebaseRulesParser#functionBody.
	exitFunctionBody(ctx) {
	}


	// Enter a parse tree produced by FirebaseRulesParser#returnStatement.
	enterReturnStatement(ctx) {
	}

	// Exit a parse tree produced by FirebaseRulesParser#returnStatement.
	exitReturnStatement(ctx) {
	}


	// Enter a parse tree produced by FirebaseRulesParser#matchRuleDeclaration.
	enterMatchRuleDeclaration(ctx) {
	}

	// Exit a parse tree produced by FirebaseRulesParser#matchRuleDeclaration.
	exitMatchRuleDeclaration(ctx) {
	}


	// Enter a parse tree produced by FirebaseRulesParser#matchStatement.
	enterMatchStatement(ctx) {
	}

	// Exit a parse tree produced by FirebaseRulesParser#matchStatement.
	exitMatchStatement(ctx) {
	}


	// Enter a parse tree produced by FirebaseRulesParser#permissionDeclaration.
	enterPermissionDeclaration(ctx) {
	}

	// Exit a parse tree produced by FirebaseRulesParser#permissionDeclaration.
	exitPermissionDeclaration(ctx) {
	}


	// Enter a parse tree produced by FirebaseRulesParser#permissionBody.
	enterPermissionBody(ctx) {
	}

	// Exit a parse tree produced by FirebaseRulesParser#permissionBody.
	exitPermissionBody(ctx) {
	}


	// Enter a parse tree produced by FirebaseRulesParser#AdditiveExpression.
	enterAdditiveExpression(ctx) {
	}

	// Exit a parse tree produced by FirebaseRulesParser#AdditiveExpression.
	exitAdditiveExpression(ctx) {
	}


	// Enter a parse tree produced by FirebaseRulesParser#RelationalExpression.
	enterRelationalExpression(ctx) {
	}

	// Exit a parse tree produced by FirebaseRulesParser#RelationalExpression.
	exitRelationalExpression(ctx) {
	}


	// Enter a parse tree produced by FirebaseRulesParser#TernaryExpression.
	enterTernaryExpression(ctx) {
	}

	// Exit a parse tree produced by FirebaseRulesParser#TernaryExpression.
	exitTernaryExpression(ctx) {
	}


	// Enter a parse tree produced by FirebaseRulesParser#LogicalAndExpression.
	enterLogicalAndExpression(ctx) {
	}

	// Exit a parse tree produced by FirebaseRulesParser#LogicalAndExpression.
	exitLogicalAndExpression(ctx) {
	}


	// Enter a parse tree produced by FirebaseRulesParser#PrimaryExpression.
	enterPrimaryExpression(ctx) {
	}

	// Exit a parse tree produced by FirebaseRulesParser#PrimaryExpression.
	exitPrimaryExpression(ctx) {
	}


	// Enter a parse tree produced by FirebaseRulesParser#LogicalOrExpression.
	enterLogicalOrExpression(ctx) {
	}

	// Exit a parse tree produced by FirebaseRulesParser#LogicalOrExpression.
	exitLogicalOrExpression(ctx) {
	}


	// Enter a parse tree produced by FirebaseRulesParser#ExprIsTypeIdExpression.
	enterExprIsTypeIdExpression(ctx) {
	}

	// Exit a parse tree produced by FirebaseRulesParser#ExprIsTypeIdExpression.
	exitExprIsTypeIdExpression(ctx) {
	}


	// Enter a parse tree produced by FirebaseRulesParser#NotExpression.
	enterNotExpression(ctx) {
	}

	// Exit a parse tree produced by FirebaseRulesParser#NotExpression.
	exitNotExpression(ctx) {
	}


	// Enter a parse tree produced by FirebaseRulesParser#UnaryAdditiveExpression.
	enterUnaryAdditiveExpression(ctx) {
	}

	// Exit a parse tree produced by FirebaseRulesParser#UnaryAdditiveExpression.
	exitUnaryAdditiveExpression(ctx) {
	}


	// Enter a parse tree produced by FirebaseRulesParser#UnsupportedAdditivePostfix.
	enterUnsupportedAdditivePostfix(ctx) {
	}

	// Exit a parse tree produced by FirebaseRulesParser#UnsupportedAdditivePostfix.
	exitUnsupportedAdditivePostfix(ctx) {
	}


	// Enter a parse tree produced by FirebaseRulesParser#ExprInExpression.
	enterExprInExpression(ctx) {
	}

	// Exit a parse tree produced by FirebaseRulesParser#ExprInExpression.
	exitExprInExpression(ctx) {
	}


	// Enter a parse tree produced by FirebaseRulesParser#UnsupportedAdditivePrefix.
	enterUnsupportedAdditivePrefix(ctx) {
	}

	// Exit a parse tree produced by FirebaseRulesParser#UnsupportedAdditivePrefix.
	exitUnsupportedAdditivePrefix(ctx) {
	}


	// Enter a parse tree produced by FirebaseRulesParser#EqualityExpression.
	enterEqualityExpression(ctx) {
	}

	// Exit a parse tree produced by FirebaseRulesParser#EqualityExpression.
	exitEqualityExpression(ctx) {
	}


	// Enter a parse tree produced by FirebaseRulesParser#MultiplicativeExpression.
	enterMultiplicativeExpression(ctx) {
	}

	// Exit a parse tree produced by FirebaseRulesParser#MultiplicativeExpression.
	exitMultiplicativeExpression(ctx) {
	}


	// Enter a parse tree produced by FirebaseRulesParser#MemberLookupSimpleExpression.
	enterMemberLookupSimpleExpression(ctx) {
	}

	// Exit a parse tree produced by FirebaseRulesParser#MemberLookupSimpleExpression.
	exitMemberLookupSimpleExpression(ctx) {
	}


	// Enter a parse tree produced by FirebaseRulesParser#IdMemberFunctionCallSimpleExpression.
	enterIdMemberFunctionCallSimpleExpression(ctx) {
	}

	// Exit a parse tree produced by FirebaseRulesParser#IdMemberFunctionCallSimpleExpression.
	exitIdMemberFunctionCallSimpleExpression(ctx) {
	}


	// Enter a parse tree produced by FirebaseRulesParser#LiteralExpression.
	enterLiteralExpression(ctx) {
	}

	// Exit a parse tree produced by FirebaseRulesParser#LiteralExpression.
	exitLiteralExpression(ctx) {
	}


	// Enter a parse tree produced by FirebaseRulesParser#ListLookupSimpleExpression.
	enterListLookupSimpleExpression(ctx) {
	}

	// Exit a parse tree produced by FirebaseRulesParser#ListLookupSimpleExpression.
	exitListLookupSimpleExpression(ctx) {
	}


	// Enter a parse tree produced by FirebaseRulesParser#MapSimpleExpression.
	enterMapSimpleExpression(ctx) {
	}

	// Exit a parse tree produced by FirebaseRulesParser#MapSimpleExpression.
	exitMapSimpleExpression(ctx) {
	}


	// Enter a parse tree produced by FirebaseRulesParser#FunctionCallSimpleExpression.
	enterFunctionCallSimpleExpression(ctx) {
	}

	// Exit a parse tree produced by FirebaseRulesParser#FunctionCallSimpleExpression.
	exitFunctionCallSimpleExpression(ctx) {
	}


	// Enter a parse tree produced by FirebaseRulesParser#MemberFunctionCallSimpleExpression.
	enterMemberFunctionCallSimpleExpression(ctx) {
	}

	// Exit a parse tree produced by FirebaseRulesParser#MemberFunctionCallSimpleExpression.
	exitMemberFunctionCallSimpleExpression(ctx) {
	}


	// Enter a parse tree produced by FirebaseRulesParser#ListSimpleExpression.
	enterListSimpleExpression(ctx) {
	}

	// Exit a parse tree produced by FirebaseRulesParser#ListSimpleExpression.
	exitListSimpleExpression(ctx) {
	}


	// Enter a parse tree produced by FirebaseRulesParser#ParenSimpleExpression.
	enterParenSimpleExpression(ctx) {
	}

	// Exit a parse tree produced by FirebaseRulesParser#ParenSimpleExpression.
	exitParenSimpleExpression(ctx) {
	}


	// Enter a parse tree produced by FirebaseRulesParser#PathSimpleExpression.
	enterPathSimpleExpression(ctx) {
	}

	// Exit a parse tree produced by FirebaseRulesParser#PathSimpleExpression.
	exitPathSimpleExpression(ctx) {
	}


	// Enter a parse tree produced by FirebaseRulesParser#VariableSimpleExpression.
	enterVariableSimpleExpression(ctx) {
	}

	// Exit a parse tree produced by FirebaseRulesParser#VariableSimpleExpression.
	exitVariableSimpleExpression(ctx) {
	}


	// Enter a parse tree produced by FirebaseRulesParser#pathExpression.
	enterPathExpression(ctx) {
	}

	// Exit a parse tree produced by FirebaseRulesParser#pathExpression.
	exitPathExpression(ctx) {
	}


	// Enter a parse tree produced by FirebaseRulesParser#pathExpressionSegment.
	enterPathExpressionSegment(ctx) {
	}

	// Exit a parse tree produced by FirebaseRulesParser#pathExpressionSegment.
	exitPathExpressionSegment(ctx) {
	}


	// Enter a parse tree produced by FirebaseRulesParser#RangeLookup.
	enterRangeLookup(ctx) {
	}

	// Exit a parse tree produced by FirebaseRulesParser#RangeLookup.
	exitRangeLookup(ctx) {
	}


	// Enter a parse tree produced by FirebaseRulesParser#IndexLookup.
	enterIndexLookup(ctx) {
	}

	// Exit a parse tree produced by FirebaseRulesParser#IndexLookup.
	exitIndexLookup(ctx) {
	}


	// Enter a parse tree produced by FirebaseRulesParser#mapExpression.
	enterMapExpression(ctx) {
	}

	// Exit a parse tree produced by FirebaseRulesParser#mapExpression.
	exitMapExpression(ctx) {
	}


	// Enter a parse tree produced by FirebaseRulesParser#mapEntries.
	enterMapEntries(ctx) {
	}

	// Exit a parse tree produced by FirebaseRulesParser#mapEntries.
	exitMapEntries(ctx) {
	}


	// Enter a parse tree produced by FirebaseRulesParser#mapEntry.
	enterMapEntry(ctx) {
	}

	// Exit a parse tree produced by FirebaseRulesParser#mapEntry.
	exitMapEntry(ctx) {
	}


	// Enter a parse tree produced by FirebaseRulesParser#expressionList.
	enterExpressionList(ctx) {
	}

	// Exit a parse tree produced by FirebaseRulesParser#expressionList.
	exitExpressionList(ctx) {
	}


	// Enter a parse tree produced by FirebaseRulesParser#literal.
	enterLiteral(ctx) {
	}

	// Exit a parse tree produced by FirebaseRulesParser#literal.
	exitLiteral(ctx) {
	}


	// Enter a parse tree produced by FirebaseRulesParser#keywordId.
	enterKeywordId(ctx) {
	}

	// Exit a parse tree produced by FirebaseRulesParser#keywordId.
	exitKeywordId(ctx) {
	}


	// Enter a parse tree produced by FirebaseRulesParser#keyword.
	enterKeyword(ctx) {
	}

	// Exit a parse tree produced by FirebaseRulesParser#keyword.
	exitKeyword(ctx) {
	}


	// Enter a parse tree produced by FirebaseRulesParser#pathDecl.
	enterPathDecl(ctx) {
	}

	// Exit a parse tree produced by FirebaseRulesParser#pathDecl.
	exitPathDecl(ctx) {
	}


	// Enter a parse tree produced by FirebaseRulesParser#SimpleSegmemnt.
	enterSimpleSegmemnt(ctx) {
	}

	// Exit a parse tree produced by FirebaseRulesParser#SimpleSegmemnt.
	exitSimpleSegmemnt(ctx) {
	}


	// Enter a parse tree produced by FirebaseRulesParser#CaptureSegment.
	enterCaptureSegment(ctx) {
	}

	// Exit a parse tree produced by FirebaseRulesParser#CaptureSegment.
	exitCaptureSegment(ctx) {
	}


	// Enter a parse tree produced by FirebaseRulesParser#GlobSegment.
	enterGlobSegment(ctx) {
	}

	// Exit a parse tree produced by FirebaseRulesParser#GlobSegment.
	exitGlobSegment(ctx) {
	}


	// Enter a parse tree produced by FirebaseRulesParser#EmptySegment.
	enterEmptySegment(ctx) {
	}

	// Exit a parse tree produced by FirebaseRulesParser#EmptySegment.
	exitEmptySegment(ctx) {
	}


	// Enter a parse tree produced by FirebaseRulesParser#globCapture.
	enterGlobCapture(ctx) {
	}

	// Exit a parse tree produced by FirebaseRulesParser#globCapture.
	exitGlobCapture(ctx) {
	}


	// Enter a parse tree produced by FirebaseRulesParser#capture.
	enterCapture(ctx) {
	}

	// Exit a parse tree produced by FirebaseRulesParser#capture.
	exitCapture(ctx) {
	}


	// Enter a parse tree produced by FirebaseRulesParser#id.
	enterId(ctx) {
	}

	// Exit a parse tree produced by FirebaseRulesParser#id.
	exitId(ctx) {
	}


	// Enter a parse tree produced by FirebaseRulesParser#serviceId.
	enterServiceId(ctx) {
	}

	// Exit a parse tree produced by FirebaseRulesParser#serviceId.
	exitServiceId(ctx) {
	}


	// Enter a parse tree produced by FirebaseRulesParser#isOperatorId.
	enterIsOperatorId(ctx) {
	}

	// Exit a parse tree produced by FirebaseRulesParser#isOperatorId.
	exitIsOperatorId(ctx) {
	}


	// Enter a parse tree produced by FirebaseRulesParser#operationId.
	enterOperationId(ctx) {
	}

	// Exit a parse tree produced by FirebaseRulesParser#operationId.
	exitOperationId(ctx) {
	}


	// Enter a parse tree produced by FirebaseRulesParser#packageId.
	enterPackageId(ctx) {
	}

	// Exit a parse tree produced by FirebaseRulesParser#packageId.
	exitPackageId(ctx) {
	}


	// Enter a parse tree produced by FirebaseRulesParser#memberId.
	enterMemberId(ctx) {
	}

	// Exit a parse tree produced by FirebaseRulesParser#memberId.
	exitMemberId(ctx) {
	}


	// Enter a parse tree produced by FirebaseRulesParser#as.
	enterAs(ctx) {
	}

	// Exit a parse tree produced by FirebaseRulesParser#as.
	exitAs(ctx) {
	}


	// Enter a parse tree produced by FirebaseRulesParser#localVariableId.
	enterLocalVariableId(ctx) {
	}

	// Exit a parse tree produced by FirebaseRulesParser#localVariableId.
	exitLocalVariableId(ctx) {
	}


	// Enter a parse tree produced by FirebaseRulesParser#localFunctionId.
	enterLocalFunctionId(ctx) {
	}

	// Exit a parse tree produced by FirebaseRulesParser#localFunctionId.
	exitLocalFunctionId(ctx) {
	}



}
